TestingScripts/CommonTasks/Components.py

```diff
@@ -0,0 +1,175 @@
+                return None
+            base_path = os.path.join('%OutRoot%Winx64', 'Product')
+            exe_path = base_path
+            if 'special_path' in Components[comp].keys():
+                for path in Components[comp]['special_path']:
+                    exe_path = os.path.join(exe_path, path)
+            else:
+                exe_path = os.path.join(exe_path, comp+"-GTest")
+            exe_path = os.path.join(exe_path, exeName+'.exe')
+            return exe_path
+
+def IgnorePathForComp(compToFind):
+    ignore = IgnoreForComp(compToFind)
+    if ignore is None:
+        return None
+    ignorePath = os.path.join(os.getenv('SrcRoot'), 'imodel-native')
+    for path in ignore:
+        ignorePath = os.path.join(ignorePath, path)
+    ignorePath = os.path.join(ignorePath, 'ignore_list.txt')
+    return ignorePath
+
+def PdbPathForComp(compToFind):
+    pdb = PdbForComp(compToFind)
+    pdbRoot = os.path.join(os.getenv('OutRoot'), 'Winx64', 'build')
+    dll = DllForComp(compToFind)
+    if dll == 'iTwinECObjects.dll':
+        pdbName = 'ECObjects'
+    if dll == 'iTwinECPresentation.dll':
+       pdbName = 'objects'
+    if dll == 'iTwinGeom.dll':
+       pdbName = 'BentleyGeom'
+    if dll == 'iTwinGeoCoord.dll':
+       pdbName = 'NonPublishedGeoCoord'
+    else:
+        pdbName = dll[:-4] #omit .dll
+        
+    if pdb is None:
+        pdbPath = os.path.join(pdbRoot, compToFind, pdbName+'.pdb')
+    else:
+        pdbPath = pdbRoot
+        for path in pdb:
+            pdbPath = os.path.join(pdbPath, path)
+        pdbPath = os.path.join(pdbPath, pdbName+'.pdb')
+    return pdbPath
+
+def RepoPathForComp(compToFind):
+    repoPath = RepoForComp(compToFind)
+    srcRoot = os.getenv('SrcRoot')
+    if repoPath is None:
+        repoPath = os.path.join(srcRoot,'imodel-native', 'iModelCore' ,compToFind)
+    return repoPath
+
+def CompForRepo(path):
+    pathToFind = os.path.join(os.getenv('SrcRoot'), 'imodel-native', path)
+    for comp in Components:
+        repoPath = RepoForComp(comp)
+        if repoPath.lower() in pathToFind.lower():
+            return comp
+    return ""
+
+def MapPathForComp(compToFind):
+    for comp in Components:
+        if compToFind.lower() in Components[comp]['product'].lower():
+            if 'map_repo' in Components[comp].keys():
+                repo = Components[comp]['map_repo']
+                repoPath = os.getenv('SrcRoot')
+                for path in repo:
+                    repoPath = os.path.join(repoPath, path)
+                return repoPath
+            else:
+                repoPath = RepoPathForComp(compToFind)
+                return repoPath
+def AllDlls():
+    dlls = []
+    for comp in Components:
+        if Components[comp]['dll'] not in dlls:
+            dlls.append(Components[comp]['dll'].lower())
+    return dlls
+def AllComps():
+    comps = []
+    for comp in Components:
+        if comp not in comps:
+            comps.append(comp.lower())
```

Comment:
each component name is different so I am just storing them in lower case because if you see my entire code everywhere all the checks and other stuff I have done on lower case

Reply:
yeah that seems reasonable. My point is that the if statement seems to intend to not add duplicate components. Why is the if statement necessary? are you worried about adding the same component twice to the comps array ?